File: lc/ArraysAndStrings/A13ProductWithoutSelf.java
package lc.ArraysAndStrings;

import lc.utils.ArrayUtils;

/**
 * https://leetcode.com/problems/product-of-array-except-self/description/
 */
public class A13ProductWithoutSelf {
    public static void main(String[] args) {
        A12Solution solution = new A12Solution();
        assert ArrayUtils.isSame1DArray(solution.productExceptSelf(new int[] {1,2,3,4}), new int[] {24,12,8,6});
        assert ArrayUtils.isSame1DArray(solution.productExceptSelf(new int[] {-1,1,0,-3,3}), new int[] {0,0,9,0,0});
    }
}

class A12Solution {
    public int[] productExceptSelf(int[] nums) {
        int overallProduct = 1;
        int zeroCount = 0;
        for(int num: nums) {
            overallProduct *= num == 0 ? 1 : num;
            if(num == 0) {
                zeroCount++;
            }
        }
        int[] result = new int[nums.length];
        if(zeroCount > 1) {
            return result;
        }
        for(int i=0;i<nums.length;i++) {
            if(zeroCount == 1 && nums[i] != 0) {
                continue;
            }
            result[i] = overallProduct/(nums[i] == 0 ? 1 : nums[i]);
        }
        return result;
    }
}

File: lc/ArraysAndStrings/A13ProductWithoutSelf.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
  const forward = [nums[0]];
  const backward = [nums[nums.length-1]];
  for(let i=1;i<nums.length;i++) {
    forward.push(forward[i-1]*nums[i]);
  }
  for(let i=nums.length-2;i>=0;i--) {
    backward.unshift(backward[0]*nums[i]);
  }
  console.log(forward, backward);
  const result = [];
  result.push(backward[1]);
  for(let i=1;i<nums.length-1;i++) {
    result.push(forward[i-1]*backward[i+1]);
  }
  result.push(forward[nums.length-2]);
  return result;
};

console.log(productExceptSelf([1, 2, 3, 4]));
console.log(productExceptSelf([-1,1,0,-3,3]));
